src/compiler/scala/tools/nsc/settings/Warnings.scala

```diff
@@ -74,6 +74,7 @@ trait Warnings {
     val Inaccessible           = LintWarning("inaccessible",              "Warn about inaccessible types in method signatures.",                       true)
     val NullaryOverride        = LintWarning("nullary-override",          "Warn when non-nullary `def f()' overrides nullary `def f'.",                true)
     val InferAny               = LintWarning("infer-any",                 "Warn when a type argument is inferred to be `Any`.",                        true)
+    val InferProductWithSerializable = LintWarning("infer-product-with-serializable", "Warn when a type argument is inferred to be `Product with Serializable`.")
```

Comment:
This does not do what it says it does.

```
> scala -Xlint:infer-any -Xlint:infer-product-with-serializable
[info] Running scala.tools.nsc.MainGenericRunner -usejavacp -Xlint:infer-any -Xlint:infer-product-with-serializable
Welcome to Scala 2.12.3-20170710-215443-d31a518 (OpenJDK 64-Bit Server VM, Java 1.8.0_131).
Type in expressions for evaluation. Or try :help.

scala> def f[A](as: A*) = 12
f: [A](as: A*)Int

scala> f(1, "2")
<console>:13: warning: a type was inferred to be `Any`; this may indicate a programming error.
       f(1, "2")
       ^
res0: Int = 12

scala> f((1,2), List(3, 4))
res1: Int = 12

scala> 
```

Instead it does (as the test cases say):

```
scala> val f = if (true) (1,2) else List(3, 4)
<console>:11: warning: a type was inferred to be `Product with Serializable`; this may indicate a programming error
       val f = if (true) (1,2) else List(3, 4)
           ^
f: Product with Serializable = (1,2)

scala> 
```

which is :+1: :+1:, but personally I think it should do both of the above.